UpstreamDownstreamDialog.py

```diff
@@ -0,0 +1,158 @@
+
+
+DOWNSTREAM_URL = "http://gs.niwa.co.nz/niwa/wms?viewparams=hydroid:%s&layers=rec2_trace_downstream&styles=&IgnoreGetMapUrl=1"
+UPSTREAM_URL = "http://gs.niwa.co.nz/niwa/wms?viewparams=hydroid:%s&layers=rec2_trace_upstream&styles=&IgnoreGetMapUrl=1"
+
+
+class PointSelectTool(QgsMapTool):
+
+    pointSelected = pyqtSignal(QgsPoint)
+
+    def __init__(self, iface):
+        QgsMapTool.__init__(self, iface.mapCanvas())
+        self.iface = iface
+
+    def canvasReleaseEvent(self, e):
+        if e.button() == Qt.LeftButton:
+            p = self.iface.mapCanvas().getCoordinateTransform().toMapCoordinates(e.pos())
+            self.pointSelected.emit(p)
+
+
+class UpstreamDownstreamDialog(QDialog, Ui_UpstreamDownstreamDialog):
+    def __init__(self, iface, recLayer, parent=None):
+        QDialog.__init__(self, parent)
+        self.setupUi(self)
+        self.iface = iface
+        self.recLayer = recLayer
+        setupDialog = SetupDialog(self.iface.mainWindow())
+
+        reg = QgsMapLayerRegistry.instance()
+        for layer in QgsMapLayerRegistry.instance().mapLayers().values():
+            if(layer.source() == "url=%s&crs=EPSG:2193&format=image/png" % setupDialog.getRecLayerURL()):
+                self.recLayer = reg.mapLayer(layer.id())
+
+        self.selectTool = PointSelectTool(self.iface)
+        self.selectTool.pointSelected.connect(
+            lambda p: self.lineEditPosition.setText("%f %f" % (p.x(), p.y())))
+
+        self.lineEditPosition.setValidator(QRegExpValidator(QRegExp("^([-+]?[0-9]*\.?[0-9]+)\s+([-+]?[0-9]*\.?[0-9]+)$")))
+        self.lineEditPosition.textChanged.connect(
+            lambda: self.pushButtonSearchDownstream.setEnabled(self.lineEditPosition.hasAcceptableInput()))
+        self.lineEditPosition.textChanged.connect(
+            lambda: self.pushButtonSearchUpstream.setEnabled(self.lineEditPosition.hasAcceptableInput()))
+
+        self.toolButtonSelectPosition.setIcon(QIcon(":/images/themes/default/mActionSelect.svg"))
+        self.toolButtonSelectPosition.clicked.connect(
+            lambda: self.iface.mapCanvas().setMapTool(self.selectTool))
+
+        self.pushButtonSearchDownstream.clicked.connect(
+            lambda: self.__addLayer(DOWNSTREAM_URL, "Downstream"))
+        self.pushButtonSearchUpstream.clicked.connect(
+            lambda: self.__addLayer(UPSTREAM_URL, "Upstream"))
+
+        self.pushButtonSetup.clicked.connect(self.__showSetupDialog)
+        self.toolButtonSelectPosition.setEnabled(False)
+        for layer in QgsMapLayerRegistry.instance().mapLayers().values():
+            if(layer.source() == "url=%s&crs=EPSG:2193&format=image/png" % setupDialog.getRecLayerURL()):
+                self.toolButtonSelectPosition.setEnabled(True)
+        QgsMapLayerRegistry.instance().layersRemoved["QStringList"].connect(self.RemoveScenario)
+
+    def RemoveScenario(self):
+        setupDialog = SetupDialog(self.iface.mainWindow())
+        self.recLayer = None
```

Comment:
I don't see how this is supposed to work. You clear the recLayer regardless of which layer was actually removed, you should check if the rec layer was the one removed. Something like

    QgsMapLayerRegistry.instance().layerWillBeRemoved.connect(self.__checkRecLayerRemoved)

    def __checkRecLayerRemoved(self, layer):
        if layer.source() == "...":
           # etc